File: EnchantsOrder/EnchantsOrder.Demo/Models/Enchantment.cs
using EnchantsOrder.Common;
using EnchantsOrder.Models;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace EnchantsOrder.Demo.Models
{
    internal class Enchantment : IEnchantment
    {
        public int Level { get; set; }
        public int Weight { get; set; }

        public bool Hidden { get; set; } = false;

        public string Name { get; set; }

        public string[] Items { get; set; }
        public string[] Incompatible { get; set; }

        public long Experience => (long)Level * Weight;

        public Enchantment(JsonProperty token)
        {
            Name = token.Name;

            JsonElement value = token.Value;
            if (value.TryGetProperty("levelMax", out JsonElement levelMaxProperty) && levelMaxProperty.TryGetInt32(out int levelMax))
            {
                Level = levelMax;
            }

            if (value.TryGetProperty("weight", out JsonElement weightProperty) && weightProperty.TryGetInt32(out int weight))
            {
                Weight = weight;
            }

            if (value.TryGetProperty("hidden", out JsonElement hiddenProperty))
            {
                Hidden = hiddenProperty.GetBoolean();
            }

            if (value.TryGetProperty("items", out JsonElement itemsProperty))
            {
                Items = itemsProperty.EnumerateArray().Select((x) => x.GetString()).ToArray();
            }

            if (value.TryGetProperty("incompatible", out JsonElement incompatibleProperty))
            {
                Incompatible = incompatibleProperty.EnumerateArray().Select((x) => x.GetString()).ToArray();
            }
        }

        /// <inheritdoc/>
        public override string ToString() => $"{Name} {Level.GetRomanNumber()}";

        /// <inheritdoc/>
        public int CompareTo(IEnchantment other)
        {
            if (other is null) { return -1; }
            int value = Experience.CompareTo(other.Experience);
            if (value == 0)
            {
                value = Level.CompareTo(other.Level);
                if (value == 0)
                {
                    value = Name.CompareTo(other.Name);
                }
            }
            return value;
        }

        /// <inheritdoc/>
        public bool Equals(IEnchantment other) => CompareTo(other) == 0;
    }
}

File: EnchantsOrder/EnchantsOrder.Web/Site/helpers/enchantment.ts
import enchants_zh from "../../../EnchantsOrder.Demo/Assets/Enchants/Enchants.zh-CN.json";
import enchants_en from "../../../EnchantsOrder.Demo/Assets/Enchants/Enchants.en-US.json";
const enchants = {
    "zh-CN": enchants_zh,
    "en-US": enchants_en
};
type Enchant = {
    hidden?: boolean,
    levelMax: number,
    weight: number,
    incompatible?: string[],
    items: string[]
}
export function getEnchantments(locale: keyof typeof enchants) {
    const json: { [key: string]: Enchant } = enchants[locale] || enchants["en-US"];
    const enchantments: Enchantment[] = [];
    for (const name in json) {
        const enchantment = json[name];
        enchantments.push({
            name,
            level: enchantment.levelMax,
            weight: enchantment.weight,
            hidden: enchantment.hidden,
            items: enchantment.items,
            incompatible: enchantment.incompatible,
            get experience() {
                return this.level * this.weight;
            }
        });
    }
    return enchantments;
}
export interface IEnchant {
    readonly level: number,
    readonly weight: number,
    readonly name: string
};
export type Enchantment = {
    name: string;
    level: number;
    weight: number;
    hidden?: boolean;
    items: string[];
    incompatible?: string[];
    readonly experience: number;
}